components/authentication-framework/org.wso2.carbon.identity.application.authentication.endpoint/src/main/webapp/add-security-questions.jsp

```diff
@@ -0,0 +1,157 @@
+
+<%@ page import="org.wso2.carbon.identity.application.authentication.endpoint.util.AuthenticationEndpointUtil" %>
+<%@ page import="org.wso2.carbon.identity.application.authentication.endpoint.util.EncodedControl" %>
+<%@ page import="org.wso2.carbon.identity.application.authentication.endpoint.util.Constants" %>
+<%@ page import="org.wso2.carbon.identity.recovery.ui.IdentityManagementAdminClient" %>
+<%@ page import="org.wso2.carbon.identity.recovery.model.ChallengeQuestion" %>
+<%@ page import="org.wso2.carbon.utils.ServerConstants" %>
+<%@ page import="org.wso2.carbon.ui.CarbonUIUtil" %>
+<%@ page import="org.wso2.carbon.CarbonConstants" %>
+
+<%@ page import="java.util.ResourceBundle" %>
+<%@ page import="java.nio.charset.StandardCharsets" %>
+
+<%@ page import="java.util.ArrayList" %>
+<%@ page import="java.util.Iterator" %>
+<%@ page import="java.util.HashMap" %>
+<%@ page import="java.util.List" %>
+<%@ page import="java.util.Map" %>
+
+<%
+    String BUNDLE = "org.wso2.carbon.identity.application.authentication.endpoint.i18n.Resources";
+    ResourceBundle resourceBundle = ResourceBundle.getBundle(BUNDLE, request.getLocale(), new
+            EncodedControl(StandardCharsets.UTF_8.toString()));
+    String urlData = request.getParameter("data");
+    // Extract the challenge questions from the request and add them into an array
+    String[] questionSets = urlData.split("&");
+    // Hash-map to hold user's challenge questions
+    Map<String, List<ChallengeQuestion>> challengeQuestionMap = new HashMap<>();
+
+    for (String question : questionSets) {
+        String[] questionProperties = question.split("\\|");
+        // Construct a new ChallengeQuestion for each challenge question received from the request
+        ChallengeQuestion tempChallengeQuestions = new ChallengeQuestion();
+        tempChallengeQuestions.setQuestionSetId(questionProperties[0]);
+        tempChallengeQuestions.setQuestionId(questionProperties[1]);
+        tempChallengeQuestions.setQuestion(questionProperties[2]);
+        // Add the challenge question to the Hash-map
+        List<ChallengeQuestion> challengeQuestionList = challengeQuestionMap.get(questionProperties[0]);
+        if (challengeQuestionList == null) {
+            List<ChallengeQuestion> tempList = new ArrayList<>();
+            tempList.add(tempChallengeQuestions);
+            challengeQuestionMap.put(questionProperties[0], tempList);
+        } else {
+            challengeQuestionList.add(tempChallengeQuestions);
+        }
+    }
+%>
+
+<script type="text/javascript">
+    function validate() {
+        var isValid = new Boolean(true);
+        $("#profile input").each(function () {
+            if ($(this).val() == "") {
+                isValid = false;
+            }
+        })
+        if (isValid) {
+            document.getElementById("profile").submit();
+        }
+        else {
+            alert("Please enter the answers for challenge questions");
```

Comment:
use CarbonUIMessage

Reply:
This jsp is inside the authentication endpoint, removed alert and used document.write to alert the user

Reply:
the page will only display that the user did not provide the required challenge questions.